torchci/components/LogViewer.tsx

```diff
@@ -1,202 +1,43 @@
-        [/\[33;1m.*?\[0m/, "yellow"],
-        [/\[34;1m.*?\[0m/, "blue"],
-        [/\[35;1m.*?\[0m/, "magenta"],
-        [/\[36;1m.*?\[0m/, "cyan"],
-        [/\[37;1m.*?\[0m/, "white"],
-
-        [/\[30m.*?\[0m/, "black"],
-        [/\[31m.*?\[0m/, "red"],
-        [/\[32m.*?\[0m/, "green"],
-        [/\[33m.*?\[0m/, "yellow"],
-        [/\[34m.*?\[0m/, "blue"],
-        [/\[35m.*?\[0m/, "magenta"],
-        [/\[36m.*?\[0m/, "cyan"],
-        [/\[37m.*?\[0m/, "white"],
-
-        [/\[0;1;30m.*?\[0m/, "black"],
-        [/\[0;1;31m.*?\[0m/, "red"],
-        [/\[0;1;32m.*?\[0m/, "green"],
-        [/\[0;1;33m.*?\[0m/, "yellow"],
-        [/\[0;1;34m.*?\[0m/, "blue"],
-        [/\[0;1;35m.*?\[0m/, "magenta"],
-        [/\[0;1;36m.*?\[0m/, "cyan"],
-        [/\[0;1;37m.*?\[0m/, "white"],
-
-        [/\[1m.*?\[0m/, "cyan"],
-        [/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{7}Z/, "hidden"],
-      ],
-    },
-  });
-
-  // Define a new theme that contains only rules that match this language
-  monaco.editor.defineTheme("logTheme", {
-    base: "vs-dark",
-    inherit: true,
-    rules: [
-      { token: "black", foreground: "000000", fontStyle: "bold" },
-      { token: "red", foreground: "af0000", fontStyle: "bold" },
-      { token: "green", foreground: "87d700", fontStyle: "bold" },
-      { token: "yellow", foreground: "d7d700", fontStyle: "bold" },
-      { token: "blue", foreground: "5fd7ff", fontStyle: "bold" },
-      { token: "magenta", foreground: "8700af", fontStyle: "bold" },
-      { token: "cyan", foreground: "00af87", fontStyle: "bold" },
-      { token: "white", foreground: "ffffff", fontStyle: "bold" },
-      { token: "hidden", foreground: "5f5f5f" },
-    ],
-    colors: {},
-  });
-
-  monaco.languages.registerFoldingRangeProvider("logText", {
-    provideFoldingRanges: function (model, context, token) {
-      const lines = model.getValue().split("\n");
-      const starts = [];
-      const ends = [];
-      let lineNumber = 0;
-      for (const line of lines) {
-        if (line.includes("##[group]")) {
-          starts.push(lineNumber);
-        } else if (line.includes("##[endgroup]")) {
-          ends.push(lineNumber);
-        }
-        lineNumber += 1;
-      }
-      let ranges = [];
-      for (let i = 0; i < Math.min(starts.length, ends.length); i++) {
-        ranges.push({
-          start: starts[i] + 1,
-          end: ends[i] + 1,
-          kind: monaco.languages.FoldingRangeKind.Imports,
-        });
-      }
-      return ranges;
-    },
-  });
-}
+import { oneDark } from "@codemirror/theme-one-dark";
+import _ from "lodash";
 
 const fetcher = (url: string) => fetch(url).then((res) => res.text());
 function Log({ url, line }: { url: string; line: number }) {
-  const monaco = useMonaco();
-
-  useEffect(() => {
-    monaco?.languages.typescript.javascriptDefaults.setEagerModelSync(true);
-  }, [monaco]);
   const { data } = useSWRImmutable(url, fetcher);
+  const viewer = useRef<HTMLDivElement>(null!);
```

Comment:
The null! is a bit weird; don't think I've ever seen it (although I don't do alot of typescript). Can we make the generic nullable or something?

Reply:
It means "I am initializing this to null but guaranteeing that it will never actually be null when accessed at runtime". So it's a way of avoiding making the generic nullable and having to check it all the time.